src/resources/ble_esp32.cc

```diff
@@ -697,6 +918,112 @@ PRIMITIVE(request_attribute) {
+  ByteArray* proxy = process->object_heap()->allocate_proxy();
+  if (!proxy) ALLOCATION_FAILED;
+
+  // Mark usage. When the group is unregistered, the usage is automatically
+  // decremented, but if group allocation fails, we manually call unuse().
+  BLEEventSource* ble = BLEEventSource::instance();
+  if (!ble->use()) {
+    MALLOC_FAILED;
+  }
+
+  BLEServerConfigGroup* group = _new BLEServerConfigGroup(process, ble);
+  if (!group) MALLOC_FAILED;
+
+  proxy->set_external_address(group);
+  return proxy;
+}
+
+PRIMITIVE(add_server_service) {
+  ARGS(BLEServerConfigGroup, group, Blob, uuid_blob);
+  ByteArray* proxy = process->object_heap()->allocate_proxy();
+  if (!proxy) ALLOCATION_FAILED;
+
+  ble_uuid_any_t uuid = uuid_from_blob(uuid_blob);
+
+  BLEServerServiceResource* service = group->add_service(uuid);
+  if (!service) MALLOC_FAILED;
+
+  proxy->set_external_address(service);
+  return proxy;
+}
+
+PRIMITIVE(add_server_characteristic) {
+  ARGS(BLEServerServiceResource, service, Blob, uuid_blob, int, type, Object, value);
+  ByteArray* proxy = process->object_heap()->allocate_proxy();
+  if (!proxy) ALLOCATION_FAILED;
+
+  ble_uuid_any_t uuid = uuid_from_blob(uuid_blob);
+
+  struct os_mbuf* om =null;
+  if (!object_to_mbuf(process, value, &om)) WRONG_TYPE;
+
+  BLEServerCharacteristicResource* characteristic = service->add_characteristic(uuid, type, om);
+  if (!characteristic) MALLOC_FAILED;
+
+  proxy->set_external_address(characteristic);
+  return proxy;
+}
+
+PRIMITIVE(set_characteristics_value) {
+  ARGS(BLEServerCharacteristicResource, resource, Object, value);
+
+  struct os_mbuf* om;
+  if (!object_to_mbuf(process, value, &om)) WRONG_TYPE;
+
+  resource->set_mbuf_to_send(om);
+
+  return process->program()->null_object();
+}
+
+PRIMITIVE(notify_characteristics_value) {
+  ARGS(BLEServerCharacteristicResource, resource, Object, value);
+
+  if (resource->is_notify_enabled() || resource->is_indicate_enabled()) {
+    struct os_mbuf* om;
+    if (!object_to_mbuf(process, value, &om)) WRONG_TYPE;
+
+    int err = 0;
+    if (resource->is_notify_enabled()) {
+      err = ble_gattc_notify_custom(resource->conn_handle(), resource->nimble_value_handle(), om);
+    }
+
+    if (err == 0 && resource->is_indicate_enabled()) {
+      err = ble_gattc_indicate_custom(resource->conn_handle(), resource->nimble_value_handle(), om);
+    }
+
+    if (err != ESP_OK) {
+      return Primitive::os_error(err, process);
+    }
+  }
+
+  if (resource->is_notify_enabled())
+    return process->program()->true_object();
+  else
+    return process->program()->false_object();
+
+  return process->program()->null_object();
+}
+
+PRIMITIVE(get_characteristics_value) {
+  ARGS(BLEServerCharacteristicResource, resource);
+
+  os_mbuf* mbuf = resource->mbuf_received();
+  if (mbuf == null) return process->program()->null_object();
+
+  Object* ret_val = convert_mbuf_to_heap_object(process, mbuf);
+
+  resource->set_mbuf_received(null);
```

Comment:
What happens if the allocation in convert_mbuf_to_heap_object fails? I don't think you want to clear the mbuf_received in that case.

Reply:
Why not? If the allocation fails, the memory is running out. if the mbuf is not cleared, it will potentially grow with further calls pressing the memory even more.

Reply:
As an allocation failure potentially can retry the primitive, the code has been updated to not clear the mbuf.